src/git/git.go

```diff
@@ -53,6 +72,345 @@ func (g *GitVersioning) validate() error {
+
+	message := fmt.Sprintf("type: [skip]: message: Commit automatically generated by Semantic Release. The new tag is %s", newReleaseVersion)
+	commit, err := worktree.Commit(message, &git.CommitOptions{Author: signature, Committer: signature})
+	if err != nil {
+		return err
+	}
+
+	g.log.Info(colorGreen+"New commit added: %s"+colorReset, commit.String())
+	return nil
+}
+
+// push aims to push the commits to the remote repository.
+// Returns:
+// 		error: Returns an error whenever unexpected issues happen.
+func (g *GitVersioning) push() error {
+	err := g.repo.Push(&git.PushOptions{
+		Auth: &http.BasicAuth{
+			Username: g.username,
+			Password: g.password,
+		},
+		InsecureSkipTLS: true})
+	if err != nil {
+		return fmt.Errorf("error while pushing commits to remote repository due to: %s", err.Error())
+	}
+
+	return nil
+}
+
+// tagExists aims to find a given tag in the repository.
+// 		tag (string): Tag name to search.
+// Returns:
+// 		bool: True when tag found, otherwise false.
+// 		error: Returns an error whenever unexpected issues happen.
+func (g *GitVersioning) tagExists(tag string) (bool, error) {
+	res := false
+	tags, err := g.repo.TagObjects()
+	if err != nil {
+		return res, fmt.Errorf("error while getting tags due to %s", err.Error())
+	}
+
+	err = tags.ForEach(func(t *object.Tag) error {
+		if t.Name == tag {
+			res = true
+			return fmt.Errorf("tag %s already exists", tag)
+		}
+		return nil
+	})
+	if err != nil && err.Error() != fmt.Sprintf("tag %s already exists", tag) {
+		return false, fmt.Errorf("error while iterating tags due to: %s", err)
+	}
+	return res, nil
+}
+
+// setTag aims to create a given tag in the local repository.
+// Args:
+// 		tag (string): Tag name to be created.
+// Returns:
+// 		err: Error whenever unexpected issues happen.
+func (g *GitVersioning) setTag(tag string) error {
+	g.log.Info("Set tag %s", tag)
+	tagExists, err := g.tagExists(tag)
+	if err != nil {
+		return err
+	}
+	if tagExists {
+		return nil
+	}
+
+	g.log.Info("Creating tag %s", tag)
+	_, err = g.repo.CreateTag(tag, g.branchHead.Hash(), &git.CreateTagOptions{
+		Tagger: &object.Signature{
+			Name:  g.mostRecentCommit.authorName,
+			Email: g.mostRecentCommit.authorEmail,
+			When:  time.Now(),
+		},
+		Message: fmt.Sprintf("Generated by semantic-release %s", tag),
+	})
+	if err != nil {
+		return fmt.Errorf("error while creating tag due to: %s", err.Error())
+	}
+	g.log.Info("Tag %s successfully created", tag)
+	return nil
+}
+
+// pushTags puhs the tags from local to remote repository
+// Returns:
+// 		error: Returns an error whenever unexpected issues happen.
+func (g *GitVersioning) pushTags() error {
+
+	po := &git.PushOptions{
+		RemoteName: "origin",
+		Progress:   os.Stdout,
+		RefSpecs:   []config.RefSpec{config.RefSpec("refs/tags/*:refs/tags/*")},
+		Auth: &http.BasicAuth{
+			Username: g.username,
+			Password: g.password,
+		},
+		InsecureSkipTLS: true,
+	}
+	err := g.repo.Push(po)
+
+	if err == nil {
```

Comment:
In the other methods usually you have used if err !=nil { ...} should we keep the same here?  @esequielvirtuoso

Reply:
I've used this approach here to avoid nested `if` conditions.
That was an observation made by @ebbauer in a previous MR.

Reply:
@esequielvirtuoso but this one is only one line :sweat_smile:

Reply:
Damm.
You're right. :D